### ISSUE TITLE
ActiveSupport::Duration values are equals but when adding them to a Date it produces different result

### ISSUE THREAD

**Issue by vmeyet** [rails/rails#35477]
### Steps to reproduce
Short Version:
```ruby
$> (1.month + 2.days) == (2.days + 1.month)
=> true # seems right
$> Date.parse('2019/03/01') - (1.month + 2.days)
=> Wed, 30 Jan 2019 # seems right
$> Date.parse('2019/03/01') - (2.days + 1.month)
=> Sun, 27 Jan 2019 # seems right

# wait... something seems wrong
```

Reproduction:

```ruby
# frozen_string_literal: true

begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  gem 'activesupport'
end

require 'active_support/all'
require 'minitest/autorun'

# Ensure backward compatibility with minitest 4.
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

# This connection will do for database-independent bug reports.

class BugTest < Minitest::Test
  def test_active_duration__durations_are_supposedly_fully_equal
    assert(duration_1 == duration_2)
    assert(duration_1.eql?(duration_2))
    assert_equal(duration_1.hash, duration_2.hash)
  end

  def test_active_duration__when_added_they_are_no_more_the_same_value
    # (begining_of_march - (1.month + 2.days)) gives #> Wed, 30 Jan 2019
    # (begining_of_march - (2.days + 1.month)) gives #> Sun, 27 Jan 2019
    assert_equal(begining_of_march - duration_1, begining_of_march - duration_2)
  end

  private

  def duration_1
    @duration_1 ||= 1.month + 2.days
  end

  def duration_2
    @duration_2 ||= 2.days + 1.month
  end

  def begining_of_march
    @begining_of_february ||= Date.parse('2019/03/01')
  end
end
```

### Expected behavior

Even though it makes sense that the following gives different result:
```ruby
$> Date.parse('2019/03/01') - 1.month - 2.days
=> Wed, 30 Jan 2019
$> Date.parse('2019/03/01') - 2.days - 1.month
=> Sun, 27 Jan 2019
```
When the duration addition is made first (parenthesis) and produce the "same" duration it should lead to the same result.

2 choices here:
- The right way (IMO), `ActiveSupport::Duration` addition is not commutative and so values should not be equal. i.e A + B != B + A
- The other way: they both are converted into a value that is equal and when adding it to a `Date`/`DateTime`/`Time` it gives the same result

### Actual behavior
Well the very same object (according to `==`, `eql?`, `hash`) added to another object gives different result.

### System configuration
**Rails version**: 5.2.1

**Ruby version**: 2.5.3

**Comment by vmeyet** [comment]
It's been almost a month and still no answers. What do you think would be the best solution to this issue?

**Comment by rafaelfranca** [comment]
cc @pixeltrix

**Comment by vishaltelangre** [comment]
The measurement unit of an `ActiveSupport::Duration` is seconds.

```ruby
>> 1.second == 1.second
=> true
>> 1.minute == 60.seconds
=> true
>> 1.hour == 3600.seconds
=> true
>> 1.day == 86400.seconds
=> true
>> 1.week == 604800.seconds
=> true
>> 1.month == 2629746.seconds # 1/12 of a gregorian year
=> true
>> 1.year == 31556952.seconds # length of a gregorian year (365.2425 days)
=> true
```

The same unit is being used to perform the comparison of these durations. The comparison is performed against the value of `#value` attribute. The `#value` represents the value of duration in terms of seconds.

```ruby
>> 2.days.value == (2 * 86400.seconds)
=> true 
>> 2.days.to_i == 2.days.value
=> true
>> (1.month + 2.days).value
=> 2802546
>> (2.days + 1.month).value
=> 2802546
>> 1.month + 2.days == 2.days + 1.month
=> true
```

Although all of this is correct, we don't seem to totally rely on this unit of the `ActiveSupport::Duration` while using it with `Date`, `DateTime`, `Time` and `ActiveSupport::TimeWithZone`.

The current behavior exists due to the fact that sometimes we use seconds and sometimes we don't.

At this moment, the [`ActiveSupport::Duration#sum`](https://github.com/rails/rails/blob/6394f9da6938d0dccfd8c95f10b2fb7ee98ab953/activesupport/lib/active_support/duration.rb#L403-L419) method works like this.

```ruby
>> a = 1.month + 2.days
>> a.parts # {:months=>1, :days=>2}
>> b = 2.days + 1.month
>> b.parts # {:days=>2, :months=>1}
>> a == b # true
>> date = Date.parse('2019/03/01') 
>> date - a # (date.advance(months: -1)).advance(days: -2)
=> Wed, 30 Jan 2019
>> date - b # (date.advance(days: -2)).advance(months: -1)
=> Sun, 27 Jan 2019
>> (date - a) == (date - b)
>> false
```

If we apply the following patch.

```patch
diff --git a/activesupport/lib/active_support/duration.rb b/activesupport/lib/active_support/duration.rb
index a30bd11a87..c0b988e61b 100644
--- a/activesupport/lib/active_support/duration.rb
+++ b/activesupport/lib/active_support/duration.rb
@@ -406,9 +406,17 @@ def sum(sign, time = ::Time.current)
             if type == :seconds
               t.since(sign * number)
             elsif type == :minutes
-              t.since(sign * number * 60)
+              t.since(sign * number * SECONDS_PER_MINUTE)
             elsif type == :hours
-              t.since(sign * number * 3600)
+              t.since(sign * number * SECONDS_PER_HOUR)
+            elsif type == :days
+              t.since(sign * number * SECONDS_PER_DAY)
+            elsif type == :weeks
+              t.since(sign * number * SECONDS_PER_WEEK)
+            elsif type == :months
+              t.since(sign * number * SECONDS_PER_MONTH)
+            elsif type == :years
+              t.since(sign * number * SECONDS_PER_YEAR)
             else
               t.advance(type => sign * number)
             end
```
Then it will always honor the `seconds` unit and yield the expected behavior.

```ruby
>> date - a
=> Sun, 27 Jan 2019 13:30:54 UTC +00:00
>> date - b
=> Sun, 27 Jan 2019 13:30:54 UTC +00:00
>> (date - a) == (date - b)
=> true
```

Notice that in the above patch, the `Time#since` is preferred over `Time#advance` to make the computation always produce the same results using seconds.

Let me know your thoughts about this approach.

_FYI, I got around 30 failing tests in `activesupport/test` directory with this patch which expect more accurate results than the current ones._

**Comment by vmeyet** [comment]
@vishaltelangre, I believe this kind of patch might make thing worse. Because `active_support/duration` aim to be the most human friendly possible.

With the current implementation
```ruby
irb(main):001:0> Time.parse('2019/02/01') + 1.month
=> 2019-03-01 00:00:00 +0000 # March 1st
```

With your patch: 1 month is converted to  `SECONDS_PER_MONTH`  seconds (i.e 2629746 seconds)
```ruby
irb(main):002:0> Time.parse('2019/02/01') +  ActiveSupport::Duration::SECONDS_PER_MONTH
=> 2019-03-03 10:29:06 +0000 # March 3rd
```

This is highly not backward compatible and goes again the human friendly goal of the lib IMHO.


After further thought about the issue, the only right solution might be to just accept that duration sum is not commutative when parts are of variable length: year, month, week, and day (mainly due to Daylight Saving Time, leap years and even leap seconds). Note that this is also a backward incompatible change but less so.


In short, 
- `1.second + 1.month` is not equal to `1.month + 1.second`
The order the parts are summed do affect the result in a meaningful way. However when we force the evaluation into seconds (`to_i`) then we default on the usual length of those variable-length duration (i.e a month is 1/12 of a gregorian year which is 2629746 seconds).
- `1.hour + 1.minute` is equal to `1.minute + 1.hour`
commutativity for invariable-length duration sum is still OK (those might even be converted into second without loss of information).


Basically, this is the same reason why the lib use `Time#advance`  for durations other than seconds, minutes and hours.

**Comment by vmeyet** [comment]
Actually to go even further the current implementation of `ActiveSupport::Duration#parts` could be considered faulty as well (same reason as this all thread)

```ruby
[1] pry(main)> date = Date.parse('2019-01-27')
=> Sun, 27 Jan 2019
[2] pry(main)> 1.month + 1.week + 1.month == 2.month + 1.week
=> true
[3] pry(main)> date + 1.month + 1.week + 1.month == date + 2.month + 1.week
=> false
```

**Comment by pixeltrix** [comment]
Sorry for missing this - was kinda busy back when it was reported.

Basically doing date maths around the end of a month is fuzzy and depends on what month end you're  dealing with. Effectively what we do at the moment is keep the order of the parts as they're added to the duration but aggregate the values. Other ways of doing this are apply them most significant to least significant so that the order is always the same or as you said keep each individual duration part and apply them in turn - each has positives/negatives and there isn't really a 'correct' solution and I'm not sure whether the probable breakage in apps is entirely worth it.

**Comment by vmeyet** [comment]
@pixeltrix thanks for your answer on this!

This issues doesn't only occurs around end of month it also occurs around Daylight Saving Time (i.e twice a year) depending on timezone.

This is indeed definitely a backward incompatible change which may not worth it. But the current behavior is clearly break the Principle of Least Astonishment.I spent quite a few hours debugging without understanding a bug i had where the duration where "equal"

**Comment by pixeltrix** [comment]
> I personally believe that "[keeping] each individual duration part and apply them in turn" is the best approach and the only negative is see is the apps breakage (which, again, might be enough of an argument to avoid doing it).

I suspect that there may be a performance impact and a slight increase in memory usage too.

Personally, I prefer to not compose durations, e.g. use them like this

``` ruby
(Date.current - 1.month) - 2.days
```

this helps prevent unexpected behaviour.

ISTR a similar duration issue with multiplication where fixing commutativity would break associativity.
